include/zephyr/senss/senss.h

```diff
@@ -0,0 +1,384 @@
+ * @brief Sensor flag indicating if this sensor is on change reporting data.
+ *
+ * Reporting sensor data when the sensor data changes.
+ *
+ * Exclusive with  \ref SENSS_SENSOR_FLAG_REPORT_ON_EVENT
+ */
+#define SENSS_SENSOR_FLAG_REPORT_ON_CHANGE			BIT(1)
+
+
+/**
+ * @brief Sensor state.
+ *
+ */
+enum senss_sensor_state {
+	SENSS_SENOSR_STATE_NOT_READY = 1,
+	SENSS_SENOSR_STATE_READY = 2,
+};
+
+
+/**
+ * @struct senss_sensor_info
+ * @brief Sensor basic constant information
+ *
+ */
+struct senss_sensor_info {
+	/** Name of the sensor instance */
+	const char *name;
+
+	/** Friendly name of the sensor instance */
+	const char *friendly_name;
+
+	/** Vendor name of the sensor instance */
+	const char *vendor;
+
+	/** Model name of the sensor instance */
+	const char *model;
+
+	/** Sensor type */
+	int32_t type;
+
+	/** Sensor index for multiple sensors under same sensor type */
+	int32_t sensor_index;
+
+	/** Sensor flags */
+	uint32_t flags;
+
+	/** Minimal report interval in micro seconds  */
+	uint32_t minimal_interval;
+
+	/** Sensor version */
+	struct senss_sensor_version version;
+};
+
+
+
+/**
+ * @brief Initialize sensor subsystem runtime.
+ *
+ * This will allocate internal resources, such as create internal schedule
+ * task, create internal management data structures.
+ *
+ * It will also enumerate all registered sensor instances in sensor subsystem,
+ * build sensors report relational data model, if a virtual sensor's required dependency
+ * sensor type not found, will return error.
+ * Then call each sensor instances' init callback API.
+ *
+ * @return 0 on success or negative error value on failure.
+ *
+ */
+int senss_init(void);
+
+ /**
+  * @brief De-initialize sensor subsystem runtime.
+  *
+  * This will release all dynamically allocated internal resources, and call
+  * all sensor instances' deinit callback API.
+  *
+  * @return 0 on success or negative error value on failure.
+  */
+int senss_deinit(void);
+
+ /**
+  * @brief Get all supported sensor instances' information.
+  *
+  * This API just returned read only information of sensor instances,
+  * no side effect to sensor instances.
+  *
+  * @param **info For receiving sensor instances' information array pointer.
+  *
+  * @return Number of sensor instances, 0 returned if no sensors.
+  */
+int senss_get_sensors(const struct senss_sensor_info **info);
+
+
+/**
+ * @brief Open sensor instance by sensor type and sensor index.
+ *
+ * Application clients use it to open a sensor instance and get its handle.
+ * Support multiple Application clients for open same sensor instance,
+ * in this case, the returned handle will different for different clients.
+ *
+ * @param type The sensor type which need to open.
+ *
+ * @param sensor_index The index which need to open.
+ *
+ * @param *handle The opened instance handle, if failed will be set to -1.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int senss_open_sensor(
+		int type, int sensor_index,
```

Comment:
`type` isn't needed

Reply:
type is the 1st level index, and sensor_index is 2nd level index to identify one specific sensor in senss

Reply:
Right, but why is that needed, why is the index not enough? It seems confusing that I would need to remember the index and the type where the type does nothing (AFAICT), see my example below. We get the list of sensors via `senss_get_sensors()` then we just use the index of the sensor we want from that list.

Reply:
`sensor_index` is used to identify multiple sensor instances for a given sensor type, fox example we can have 3 sensor instances of same Accelerometer sensor type, we can use `sensor_index` of `0`, `1` and `2` to identify them. 

`type` + `sensor_index` is common and nature for user to uniquely identify a sensor instance, for user it's more easy to remember of a sensor `type` (user more care about `type`) instead of a global index,  for example user can use `senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 0)` to open the first or default Accelerometer sensor and use `senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 1)` to open the second Accelerometer sensor if exists.   

Global index has another problem, that sensor types and instances are not consistent for different board configurations, can't make consistent global index, it's hard for user to use. 

[chre](https://android.googlesource.com/platform/system/chre/+blame/refs/heads/master/chre_api/include/chre_api/chre/sensor.h#623) used the same idea to identify sensor instances: 

```c
/**
  * Uniquely identifies the sensor for a given type. A value of 0 indicates
  * that this is the "default" sensor, which is returned by
  * chreSensorFindDefault().
  *
  * The sensor index of a given type must be stable across boots (i.e. must
  * not change), and a different sensor of the same type must have different
  * sensor index values, and the set of sensorIndex values for a given sensor
  * type must be continuguous.
  *
  * @since v1.5
  */
  uint8_t sensorIndex;
```
and the [chreSensorFind](https://android.googlesource.com/platform/system/chre/+blame/refs/heads/master/chre_api/include/chre_api/chre/sensor.h#768)
```c
bool chreSensorFind(uint8_t sensorType, uint8_t sensorIndex, uint32_t *handle);
```

Reply:
ACK, I must have missed this addition in 1.5.

Reply:
Yes, `chreSensorFind` is the nanoapp API not the PAL. 

Since `senss`'s target application clients not just chre PAL, but also can have others, like HID client for Windows AP, and Zephyr native application client for IoT products etc.  

Especially for Zephyr native application client, `senss` API will service as final application API, not just a middle API to implement other frameworks like chre PAL.

So, `type` + `sensor_index` concept in `senss` API is more easy and nature to understand and use.

Reply:
`type` + `sensor_index` is more intuitive, I mean Application can just use `senss_open_sensor()` directly without call `senss_find_sensor()` to return a sensor info array at first.

For example, a native simple App can just use below code to access a Acc sensor:

```c
senss_open_sensor(SENSS_SENSOR_TYPE_MOTION_ACCELEROMETER_3D, 0)
```
Using `senss_find_sensor()` to enumerate all sensors array is useful for Apps which want to know what sensors in current system and dynamically use of them, but for some simple IoT Apps which has access of specific sensors, such as access a Gryscope and Acc, no need call `senss_find_sensor()` at first, can just call `senss_open_sensor(type, index)` directly, this is more intuitive.